docs: add non library solution for video page (#17094)
microsoft/playwright: docs/src/videos.md

title: "Videos"
---

With Playwright you can record videos for your tests.

> ## Record video

Reviewer: Let's do 3 tabs (js/ts/library) and merge the two sections together.

* langs: js

Playwright Test can record videos for your tests, controlled by the `video` option. By default videos are off.

- `'off'` - Do not record video.
- `'on'` - Record video for each test.
- `'retain-on-failure'` - Record video for each test, but remove all videos from successful test runs.
- `'on-first-retry'` - Record video only when retrying a test for the first time.

Video files will appear in the test output directory, typically `test-results`. See [`property: TestOptions.video`] for advanced video configuration.

```js tab=js-js
// @ts-check

/** @type {import('@playwright/test').PlaywrightTestConfig} */
const config = {
  use: {
    video: 'on-first-retry',
  },
};

module.exports = config;
```

```js tab=js-ts
import type { PlaywrightTestConfig } from '@playwright/test';
const config: PlaywrightTestConfig = {
  use: {
    video: 'on-first-retry',
  },
};
export default config;
```

> ## Library

Reviewer: ```suggestion
## On demand / manually
```

* langs: js

_For when not using Playwright as a Test Runner_

> Playwright can record videos for all pages in a [browser context](./browser-contexts.md). Videos are saved upon context closure, so make sure to await [`method: BrowserContext.close`].

Reviewer: ```suggestion
Videos are saved upon [browser context](./browser-contexts.md) closure at the end of a test. If you create a browser context manually, make sure to await [`method: BrowserContext.close`].
```

```js
const context = await browser.newContext({ recordVideo: { dir: 'videos/' } });
// Make sure to await close, so that videos are saved.
await context.close();
```

> You can also specify video size. The video size defaults to the viewport size scaled down to fit 800x800. The video of the viewport is placed in the top-left corner of the output video, scaled down to fit if necessary. You may need to set the viewport size to match your desired video size.

Reviewer: Video size also applies to the config.

```js
const context = await browser.newContext({
  recordVideo: {
    dir: 'videos/',
    size: { width: 640, height: 480 },
  }
});
```

> Saved video files will appear in the specified folder. They all have generated unique names.

Reviewer: ```suggestion
```

> For the multi-page scenarios, you can access the video file associated with the page via the

Reviewer: Keep this part below multi-tab snippets.

[`method: Page.video`].

```js
const path = await page.video().path();
```

:::note
Note that the video is only available after the page or browser context is closed.
:::

## Record video
* langs: python, java, csharp

> Playwright can record videos for all pages in a [browser context](./browser-contexts.md). Videos are saved upon context closure, so make sure to await [`method: BrowserContext.close`].

Reviewer: ```suggestion
Playwright can record videos for all pages in a [browser context](./browser-contexts.md). You must set up video recording in [`method: Browser.newContext`]. Videos are saved upon context closure, so make sure to await [`method: BrowserContext.close`].
```

> - [BrowserContext]

Reviewer: Drop the reference.
